### PR TITLE
otelgrpc: Add SpanStartOptions

### LINKED ISSUE #3683: otelgrpc: add WithSpanOptions
`otelgrpc` interceptors make spans, but those spans are not configurable. `otelhttp` has `WithSpanOptions` function which can be used to configure spans in otelhttp middleware both for transport and handler. `otelgrpc` interceptors ought to have the same functionality.

### PR DESCRIPTION
**Pull request by dubonzi**
Closes #3683.

### CODE REVIEW

File: instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:88]: This is not safe to do as this function can be called concurrently from multiple goroutines and the backing array of the `defaultStartOpts` might be modified concurrently by this `append`. This can only happen if the slice happens to have unused capacity i.e. the array is re-used rather than re-allocated, but there is a chance nonetheless. I hope this makes sense.

A solution is to explicitly allocate a new slice, copy elements from `defaultStartOpts` and append `trace.WithAttributes()`.

Comment (dubonzi) [reply instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:88]: You are correct, I didn't notice that when moving the slice creation to outside of the inner function, and that can definitely happen. My idea was to avoid calling `trace.WithSpanKind` and `cfg.SpanStartOptions...` for every span since those options never change. Do you think thats a worthwhile optmization?
If not, I can always revert to creating the whole slice for each span which would avoid this problem.

Comment (ash2k) [reply instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:88]: It probably is a minor but worthwhile optimization. The code will be allocating a slice one way or another so no harm in making it a bit more efficient even if a bit more clunky.

Comment (ash2k) [reply instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:88]: On the other hand, maybe Go compiler can optimize/inline the allocation/call of the option function and there is no point. I'll leave this to maintainers to decide.

Comment (dubonzi) [reply instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go]: Sorry for leaving this unsolved, I chose to revert the changes to avoid the problem.

File: instrumentation/google.golang.org/grpc/otelgrpc/config.go
Comment (ash2k) [inline instrumentation/google.golang.org/grpc/otelgrpc/config.go:142]: This `if` is not actually needed, right?

Everything else looks good, thanks!

Comment (dubonzi) [reply instrumentation/google.golang.org/grpc/otelgrpc/config.go]: You are correct.

**Review by ash2k — APPROVED** [review]
LGTM, but I'm not a maintainer here.

**Comment by pellared** [comment]
@dubonzi Can you please update the branch (resolve conflicts)?

@dashpole @hanyuancheung Can you please review (as codeowners)?

**Comment by dubonzi** [comment]
> @dubonzi Can you please update the branch (resolve conflicts)?
> 
> @dashpole @hanyuancheung Can you please review (as codeowners)?

All done

**Review by pellared — APPROVED** [review]

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**
